### PR TITLE
tracing: add intel ADSP memory window backend

### PR DESCRIPTION
**Pull request by marcinszkudlinski**
This commits adds a tracing backend based on Intel ADSP debug memory window

### CODE REVIEW

File: subsys/tracing/tracing_backend_adsp_memory_window.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		to_copy = length;
+
+	memcpy((void *)(mem_window->data + mem_window->head_offset), data, to_copy);
+
+	length -= to_copy;
+	if (length) {
+		memcpy((void *)mem_window->data, data + to_copy, length);
+		mem_window->head_offset = length;
+	} else
+		mem_window->head_offset += to_copy;
```
Comment (andyross) [inline subsys/tracing/tracing_backend_adsp_memory_window.c:47]: Pedantry: you need both a compiler and hardware write barrier after the data memcpy() and before the update of head_offset.  In practice the hardware requirements are not a problem on (uncached!) Xtensa addresses as the core is fully in-order.   And as long as the memcpy()'s are true function calls and not inlined via magic, the compiler can't mess that up either.  But LTO is coming, so the barrier might not be a terrible idea.

File: subsys/tracing/tracing_backend_adsp_memory_window.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		uint8_t *data, uint32_t length)
+{
+	/* copy data to ring buffer,
+	 * to make FW part fast, there's no sync with the data reader
+	 * the reader MUST read data before they got overwritten
+	 */
+	size_t to_copy = ADSP_TRACING_WINDOW_DATA_SIZE - mem_window->head_offset;
+
+	if (to_copy > length)
+		to_copy = length;
```
Comment (andyross) [inline subsys/tracing/tracing_backend_adsp_memory_window.c:38]: `to_copy = MIN(length, ADSP_TRACING_WINDOW_DATA_SIZE - mem_window->head_offset);` is a cleaner way to express this

File: subsys/tracing/tracing_core.c [hunk 3]
```diff
@@ -29,6 +29,8 @@
 #define TRACING_BACKEND_NAME "tracing_backend_posix"
 #elif defined CONFIG_TRACING_BACKEND_RAM
 #define TRACING_BACKEND_NAME "tracing_backend_ram"
+#elif defined CONFIG_TRACING_BACKEND_ADSP_MEMORY_WINDOW
+#define TRACING_BACKEND_NAME "tracing_backend_adsp_memory_window"
```
Comment (andyross) [inline subsys/tracing/tracing_core.c:33]: This is... a really weird API.  Selecting a string value in a C file based on a bunch of kconfigs?  Why not just make it a string kconfig?  Or put a name field in the backend struct.  Not the fault of this PR obviously.

**Review by andyross — DISMISSED** [review]
Very reasonable.  Some nitpicking over lockless algorithm hygiene.  I was going to whine about the lack of synchronization, but IIRC this is a design like logging where the backend is only called out of a single thread?

**Comment by marcinszkudlinski** [comment]
> I was going to whine about the lack of synchronization, but IIRC this is a design like logging where the backend is only called out of a single thread?

The sync is at Kenrel side (either a therad "async" or irq_lock "sync"), this is only a backend that does not to be reentrant:
```
void tracing_format_raw_data(uint8_t *data, uint32_t length)
{
	if (!is_tracing_enabled()) {
		return;
	}

	TRACING_LOCK();
	tracing_buffer_handle(data, length);
	TRACING_UNLOCK();
}

#define TRACING_LOCK()		{ int key; key = irq_lock()

#define TRACING_UNLOCK()	{ irq_unlock(key); } }

```

File: soc/intel/intel_adsp/common/include/adsp_debug_window.h [hunk 4]
```diff
@@ -41,12 +41,16 @@
 #define ADSP_DW_SLOT_SIZE		ADSP_DW_PAGE_SIZE
 #define ADSP_DW_SLOT_COUNT		15
 
+/* debug logs slots usage */
+#define ADSP_DW_SLOT_NUM_CTF		1
```
Comment (nashif) [inline soc/intel/intel_adsp/common/include/adsp_debug_window.h:45]: can we call this _TRACE and be format agnostic? we might want to push some different type of trace format, I do not think there is anything CTF specific about this.

**Review by kv2019i — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**